plugins/Terminal/js/components/commandinput.js

```diff
@@ -0,0 +1,75 @@
+import React from 'react'
+import { Map } from 'immutable'
+
+//These commands need a password prompt.
+const specialCommands = [ ['wallet', 'unlock'], ['wallet', 'load', 'seed'], ['help'], ['?'] ]
+
+const CommandInput = ({currentCommand, showCommandOverview, actions}) => {
+	componentDidUpdate: {
+		setTimeout(() => document.getElementById('command-input').focus(), 0)
+	}
+
+	reduce: {
+		const handleTextInput = (e) => {
+			actions.setCurrentCommand(e.target.value)
+		}
+
+		const handleKeyboardPress = (e) => {
+
+			//Enter button.
+			if (e.keyCode === 13) {
+
+				//Check if command is special.
+				switch (  isCommandSpecial(currentCommand, specialCommands) ) {
+				case -1: //Regular command.
+					spawnCommand(currentCommand, actions) //Spawn command defined in index.js.
+					break
+
+				case 0: //wallet unlock
+				case 1: //wallet load seed
+					actions.showWalletPrompt()
+					break
+
+				case 2: //help
+					var text = 'help'
+				case 3: //?
+					var newText = text || '?'
+					if (showCommandOverview) {
+						actions.hideCommandOverview()
+					} else {
+						actions.showCommandOverview()
+					}
+
+					//The command history won't actually show a help command so it is fine to add the command.
+					var newCommand = Map({ command: newText, result: '', id: Math.floor(Math.random()*1000000) })
```

Comment:
The ui updating and command dispatching is fully async and the ui currently allows multiple commands to be updating simultaneously. When a command sends output the id is used to identify and append the output to the proper object in the redux state. This prevents issues that could occur if, for example, `version` command was run twice in a row but both returned at the same time. I wanted to make sure we didn't end up with one displaying nothing and the other displaying "version 1.00version1.00" or something of the sort.

I could sequentially number commands and use this as the id but I was worried that may introduce subtle race-time issues. If react/electron don't mess with the atomic nature of js it may not be an issue. An alternative is to lock the ui input when running a command so only one command could receive output at a time. Might be an issue, ie lock up the ui, for long running commands like `wallet unlock` or if a command doesn't quit correctly.

As for adding the command, this is so that `help` and `?` appear in the command history. It isn't displayed but allows the user to hit the up arrow to pull up the command.